src/users/repositories/users.repository.ts

```diff
@@ -0,0 +1,127 @@
+import z from 'zod';
+import { v7 as uuidv7 } from 'uuid';
+
+import { db } from '../db.js';
+import { logger } from '../../common/logger.js';
+import { NewUser, User } from '../routes/types.js';
+
+const createUserCommandResultSchema = z.object({
+  id: z.string().uuid(),
+  is_inserted: z.boolean(),
+  is_updated: z.boolean(),
+  was_reactivated: z.union([z.boolean(), z.null()]),
+});
+
+const userSchema = z
+  .object({
+    email: z.string().email(),
+    full_name: z.string(),
+    joined_at: z.date(),
+  })
+  .transform((data) => ({
+    email: data.email,
+    fullName: data.full_name,
+    joinedAt:
+      data.joined_at instanceof Date
+        ? data.joined_at.toISOString()
+        : data.joined_at,
+  }));
+
+export const UsersRepository = {
+  createUser: async (user: NewUser): Promise<void> => {
+    const { email, fullName } = user;
+
+    const query = `
+        WITH existing_user AS (
+            SELECT deleted_since as previous_deleted_since
+            FROM users 
+            WHERE email = $2
+        )
+        INSERT INTO users (id, email, full_name)
+        VALUES ($1, $2, $3)
+        ON CONFLICT (email)
+        DO UPDATE SET 
+            full_name = EXCLUDED.full_name,
+            deleted_since = NULL
+        RETURNING 
+            id, 
+            (xmax = 0) as is_inserted, 
+            (xmax != 0) as is_updated,
+            (SELECT (xmax != 0 AND previous_deleted_since IS NOT NULL) FROM existing_user) as was_reactivated
+    `;
+
+    const result = await db.query(query, [uuidv7(), email, fullName]);
```

Comment:
coding style (not essential, but still): for debugging purpose, I'd extract the `uuidv7()` part to a local variable.